### PR TITLE
implementing matmul with scipy

### PR DESCRIPTION
**Pull request by joker2411**
## Description of the change

> the converge method was using numpy functions. This is very inefficient at large scale. We created a branch using torch to resolve this. But this similar functionality can be achieved through scipy, as scipy is a lot more generic than torch. Ticket [Link](https://linear.app/rudderstack/issue/PRML-648/move-attribution-model-to-scipylinalg).

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

**Review by shubhammehra4 — APPROVED** [review]

File: src/profiles-pycorelib/attribution_model.py [hunk 1]
```diff
@@ -114,36 +119,36 @@ def _plot_transitions(transition_probabilities: np.array,
...
+        neg_transitions, labels = self._generate_transition_counts(dropoff_touchpoints_list, distinct_touches_list, is_positive=False)
         all_transitions = pos_transitions + neg_transitions
         transition_probabilities = row_normalize_np_array(all_transitions)
         return transition_probabilities, labels
-    @staticmethod
-    def _converge(transition_matrix, max_iters=200, verbose=True):
-        T_upd = transition_matrix
-        prev_T = transition_matrix
+    
+    def _converge(self, transition_matrix, max_iters=40, verbose=True):
```
Comment (dpatchigolla) [inline src/profiles-pycorelib/attribution_model.py:136]: Can you make verbose=False here

File: src/profiles-pycorelib/attribution_model.py [hunk 2]
```diff
@@ -152,40 +157,50 @@ def _get_removal_affects(transition_probs, labels, ignore_labels=["Start", "Drop
                 drop_transition = transition_probs.copy()
                 drop_transition[n,:] = 0. # Drop all transitions from this touchpoint
                 drop_transition[n,-2] = 1. # Force all touches to dropoff from this touchpoint
-                drop_transition_converged = MultiTouchModels._converge(drop_transition, 500, False)
+                drop_transition_converged = self._converge(drop_transition, max_iters=500, verbose=False)
```
Comment (dpatchigolla) [inline src/profiles-pycorelib/attribution_model.py:160]: Can you remove the max_iters and verbose params here, and let the defaults be used? We don't need 500 iterations.

File: src/profiles-pycorelib/attribution_model.py [hunk 3]
```diff
@@ -152,40 +157,50 @@ def _get_removal_affects(transition_probs, labels, ignore_labels=["Start", "Drop
...
                 removal_affect[label] = default_conversion - drop_transition_converged[0,-1]
         return removal_affect
     
-    @staticmethod
-    def _get_markov_scores(tp_list_positive: List[List[Union[int, str]]],
+    def _converged_transition_probabilities(self, transition_probabilities):
+        tic = time.time()
+        converged_transition_probabilities = self._converge(transition_probabilities, max_iters=500, verbose=False)
```
Comment (dpatchigolla) [inline src/profiles-pycorelib/attribution_model.py:166]: Same as above. No need of the default params.

File: src/profiles-pycorelib/attribution_model.py [hunk 4]
```diff
@@ -56,7 +58,10 @@ def validate(self) -> Tuple[bool, str]:
         return True, "Validated successfully"
 
 class MultiTouchModels:
-    def linear_model(input_df, touchpoints_array_col, conversion_col:str) -> pd.DataFrame:
+    def __init__(self):
+        self.logger = Logger("multitouch_models") # create a logger for debug/error logging
```
Comment (dpatchigolla) [inline src/profiles-pycorelib/attribution_model.py:62]: You can pass the same `attribution_model` logger from AttributionLogger class in init and use that.

**Review by dpatchigolla — CHANGES_REQUESTED** [review]

**Review by dpatchigolla — APPROVED** [review]

**Pull request merged**